src/Converter/RefundUnitsConverter.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\RefundPlugin\Converter;
+
+use Sylius\RefundPlugin\Calculator\UnitRefundTotalCalculatorInterface;
+use Sylius\RefundPlugin\Model\RefundType;
+use Webmozart\Assert\Assert;
+
+final class RefundUnitsConverter implements RefundUnitsConverterInterface
+{
+    /** @var UnitRefundTotalCalculatorInterface */
+    private $unitRefundTotalCalculator;
+
+    public function __construct(UnitRefundTotalCalculatorInterface $unitRefundTotalCalculator)
+    {
+        $this->unitRefundTotalCalculator = $unitRefundTotalCalculator;
+    }
+
+    public function convert(array $units, RefundType $refundType, string $unitRefundClass): array
+    {
+        $units = $this->filterEmptyRefundUnits($units);
+        $refundUnits = [];
+        foreach ($units as $id => $unit) {
+            $total = $this->unitRefundTotalCalculator
+                ->calculateForUnitWithIdAndType($id, $refundType, $this->getAmount($unit))
+            ;
+
+            $refundUnits[] = new $unitRefundClass((int) $id, $total);
+        }
+
+        return $refundUnits;
+    }
+
+    private function filterEmptyRefundUnits(array $units): array
+    {
+        return array_filter($units, function (array $refundUnit): bool {
+            return (isset($refundUnit['amount']) && $refundUnit['amount'] !== '') || isset($refundUnit['full']);
+        });
+    }
+
+    private function getAmount(array $unit): ?float
+    {
+        if (isset($unit['full'])) {
+            return null;
+        }
```

Comment:
This is so bad :( I wonder, cannot we just always operate on the actual value? I don't know won't it need a avalanche of changes in other parts of the application, so could possibly be done as an improvement in a separate PR

Reply:
I agree, but I didn't want to change the current behaviour, just extract some logic and improve extendability